src/server/routes/register.ts

```diff
@@ -110,6 +133,262 @@ router.post(
+				// if it does, then we know the stateHandle is valid and we're in the correct state
+				validateIntrospectRemediation(
+					introspectResponse,
+					'enroll-authenticator',
+				);
+
+				// attempt to answer the passcode challenge, if this fails, it falls through to the catch block where we handle the error
+				const challengeAnswerResponse = await challengeAnswerPasscode(
+					stateHandle,
+					{ passcode: code },
+					requestId,
+				);
+
+				// if passcode challenge is successful, we can proceed to the next step
+				// which is to enroll a password authenticator, as we still need users to set a password for the time being
+				// we first look for the password authenticator id deep in the response
+				const passwordAuthenticatorId =
+					challengeAnswerResponse.remediation.value
+						.flatMap((remediation) => {
+							if (remediation.name === 'select-authenticator-enroll') {
+								const parsedRemediation =
+									selectAuthenticationEnrollSchema.safeParse(remediation);
+
+								if (parsedRemediation.success) {
+									return parsedRemediation.data.value.flatMap((value) => {
+										if (value.name === 'authenticator') {
+											return value.options.flatMap((option) => {
+												if (option.label === 'Password') {
+													if (
+														option.value.form.value.some(
+															(v) => v.value === 'password',
+														)
+													) {
+														return [
+															option.value.form.value.find(
+																(v) => v.name === 'id',
+															)?.value,
+														];
+													}
+												}
+											});
+										}
+									});
+								}
+							}
+						})
+						.filter(
+							(id): id is string => typeof id === 'string' && id.length > 0,
+						);
+
+				if (!passwordAuthenticatorId.length) {
+					throw new OAuthError(
+						{
+							error: 'idx_error',
+							error_description: 'Password authenticator id not found',
+						},
+						400,
+					);
+				}
+
+				// start the credential enroll flow
+				await credentialEnroll(
+					stateHandle,
+					{ id: passwordAuthenticatorId[0], methodType: 'password' },
+					requestId,
+				);
+
+				// redirect to the password page to set a password
+				return res.redirect(
+					303,
+					addQueryParamsToPath('/welcome/password', res.locals.queryParams),
+				);
+			} catch (error) {
+				if (error instanceof OAuthError) {
+					if (error.name === 'api.authn.error.PASSCODE_INVALID') {
+						// case for invalid passcode
+						const html = renderer('/register/email-sent', {
+							requestState: mergeRequestState(res.locals, {
+								queryParams: {
+									returnUrl: res.locals.queryParams.returnUrl,
+									emailSentSuccess: false,
+								},
+								pageData: {
+									email: readEncryptedStateCookie(req)?.email,
+									timeUntilTokenExpiry: encryptedState.stateHandleExpiresAt
+										? new Date(encryptedState.stateHandleExpiresAt).getTime() -
+											Date.now()
```

Comment:
I wonder if this could be a helper/lib method.

Reply:
converted this to a `convertExpiresAtToExpiryTimeInMs` method